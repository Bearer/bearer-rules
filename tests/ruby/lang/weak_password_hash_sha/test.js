const {
  createNewInvoker,
  getEnvironment,
} = require("../../../helper.js")
const { ruleId, ruleFile, testBase } = getEnvironment(__dirname)

describe(ruleId, () => {
  const invoke = createNewInvoker(ruleId, ruleFile, testBase)

  
    test("digest_sha1_data", () => {
      const testCase = "digest_sha1_data.rb"

      const results = invoke(testCase)

      expect(results.Missing).toEqual([])
      expect(results.Extra).toEqual([])
    })
  

    test("ok_not_a_password", () => {
      const testCase = "ok_not_a_password.rb"

      const results = invoke(testCase)

      expect(results.Missing).toEqual([])
      expect(results.Extra).toEqual([])
    })
  

    test("openssl_sha1_data", () => {
      const testCase = "openssl_sha1_data.rb"

      const results = invoke(testCase)

      expect(results.Missing).toEqual([])
      expect(results.Extra).toEqual([])
    })
  
})