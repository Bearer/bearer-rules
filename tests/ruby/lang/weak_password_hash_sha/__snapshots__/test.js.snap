// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ruby_lang_weak_password_hash_sha digest_sha1_data 1`] = `
"{
  "high": [
    {
      "cwe_ids": [
        "326"
      ],
      "id": "ruby_lang_weak_password_hash_sha",
      "title": "Usage of weak hashing library on a password (SHA)",
      "description": "## Description\\n\\nA weak hashing library can lead to data breaches and greater security risk.\\n\\n## Remediations\\nAccording to [OWASP](https://owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/09-Testing_for_Weak_Cryptography/04-Testing_for_Weak_Encryption), SHA and SHA1 are considered weak hash algorithms and therefore shouldn't be used.\\n\\n✅ Use a recommended hashing library such as bcrypt or Argon2id for passwords\\n\\n\`\`\`ruby\\nBCrypt::Password.create(\\"password_123\\")\\n\`\`\`\\n\\n\`\`\`ruby\\nhasher = Argon2::Password.new()\\nhasher.create(\\"password_123\\")\\n\`\`\`\\n\\n## Resources\\n- [OWASP Password Storage Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html)\\n- [BCrypt Explained](https://dev.to/sylviapap/bcrypt-explained-4k5c)\\n- [ruby-argon2](https://github.com/technion/ruby-argon2)\\n",
      "documentation_url": "https://docs.bearer.com/reference/rules/ruby_lang_weak_password_hash_sha",
      "line_number": 1,
      "full_filename": "/tmp/bearer-scan/digest_sha1_data.rb",
      "filename": ".",
      "data_type": {
        "category_uuid": "dd88aee5-9d40-4ad2-8983-0c791ddec47c",
        "name": "Passwords"
      },
      "category_groups": [
        "PII",
        "Personal Data"
      ],
      "source": {
        "start": 1,
        "end": 1,
        "column": {
          "start": 24,
          "end": 37
        }
      },
      "sink": {
        "start": 1,
        "end": 1,
        "column": {
          "start": 1,
          "end": 38
        },
        "content": "Digest::SHA1.hexdigest(user.password)"
      },
      "parent_line_number": 1,
      "snippet": "Digest::SHA1.hexdigest(user.password)",
      "fingerprint": "a917eee0d88b45489c7189f37e092168_0",
      "old_fingerprint": "582d72cdeecd01fe0be8076586b2d871_0",
      "code_extract": "Digest::SHA1.hexdigest(user.password)"
    },
    {
      "cwe_ids": [
        "326"
      ],
      "id": "ruby_lang_weak_password_hash_sha",
      "title": "Usage of weak hashing library on a password (SHA)",
      "description": "## Description\\n\\nA weak hashing library can lead to data breaches and greater security risk.\\n\\n## Remediations\\nAccording to [OWASP](https://owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/09-Testing_for_Weak_Cryptography/04-Testing_for_Weak_Encryption), SHA and SHA1 are considered weak hash algorithms and therefore shouldn't be used.\\n\\n✅ Use a recommended hashing library such as bcrypt or Argon2id for passwords\\n\\n\`\`\`ruby\\nBCrypt::Password.create(\\"password_123\\")\\n\`\`\`\\n\\n\`\`\`ruby\\nhasher = Argon2::Password.new()\\nhasher.create(\\"password_123\\")\\n\`\`\`\\n\\n## Resources\\n- [OWASP Password Storage Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html)\\n- [BCrypt Explained](https://dev.to/sylviapap/bcrypt-explained-4k5c)\\n- [ruby-argon2](https://github.com/technion/ruby-argon2)\\n",
      "documentation_url": "https://docs.bearer.com/reference/rules/ruby_lang_weak_password_hash_sha",
      "line_number": 3,
      "full_filename": "/tmp/bearer-scan/digest_sha1_data.rb",
      "filename": ".",
      "data_type": {
        "category_uuid": "dd88aee5-9d40-4ad2-8983-0c791ddec47c",
        "name": "Passwords"
      },
      "category_groups": [
        "PII",
        "Personal Data"
      ],
      "source": {
        "start": 3,
        "end": 3,
        "column": {
          "start": 25,
          "end": 38
        }
      },
      "sink": {
        "start": 3,
        "end": 3,
        "column": {
          "start": 1,
          "end": 39
        },
        "content": "Digest::SHA1.new.update(user.password)"
      },
      "parent_line_number": 3,
      "snippet": "Digest::SHA1.new.update(user.password)",
      "fingerprint": "a917eee0d88b45489c7189f37e092168_1",
      "old_fingerprint": "582d72cdeecd01fe0be8076586b2d871_1",
      "code_extract": "Digest::SHA1.new.update(user.password)"
    }
  ]
}"
`;

exports[`ruby_lang_weak_password_hash_sha ok_not_a_password 1`] = `"{}"`;

exports[`ruby_lang_weak_password_hash_sha openssl_sha1_data 1`] = `
"{
  "high": [
    {
      "cwe_ids": [
        "326"
      ],
      "id": "ruby_lang_weak_password_hash_sha",
      "title": "Usage of weak hashing library on a password (SHA)",
      "description": "## Description\\n\\nA weak hashing library can lead to data breaches and greater security risk.\\n\\n## Remediations\\nAccording to [OWASP](https://owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/09-Testing_for_Weak_Cryptography/04-Testing_for_Weak_Encryption), SHA and SHA1 are considered weak hash algorithms and therefore shouldn't be used.\\n\\n✅ Use a recommended hashing library such as bcrypt or Argon2id for passwords\\n\\n\`\`\`ruby\\nBCrypt::Password.create(\\"password_123\\")\\n\`\`\`\\n\\n\`\`\`ruby\\nhasher = Argon2::Password.new()\\nhasher.create(\\"password_123\\")\\n\`\`\`\\n\\n## Resources\\n- [OWASP Password Storage Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html)\\n- [BCrypt Explained](https://dev.to/sylviapap/bcrypt-explained-4k5c)\\n- [ruby-argon2](https://github.com/technion/ruby-argon2)\\n",
      "documentation_url": "https://docs.bearer.com/reference/rules/ruby_lang_weak_password_hash_sha",
      "line_number": 1,
      "full_filename": "/tmp/bearer-scan/openssl_sha1_data.rb",
      "filename": ".",
      "data_type": {
        "category_uuid": "dd88aee5-9d40-4ad2-8983-0c791ddec47c",
        "name": "Passwords"
      },
      "category_groups": [
        "PII",
        "Personal Data"
      ],
      "source": {
        "start": 1,
        "end": 1,
        "column": {
          "start": 32,
          "end": 45
        }
      },
      "sink": {
        "start": 1,
        "end": 1,
        "column": {
          "start": 1,
          "end": 46
        },
        "content": "OpenSSL::Digest.digest(\\"sha1\\", user.password)"
      },
      "parent_line_number": 1,
      "snippet": "OpenSSL::Digest.digest(\\"sha1\\", user.password)",
      "fingerprint": "a917eee0d88b45489c7189f37e092168_0",
      "old_fingerprint": "cd1923849b36c6435d5712d897d6da17_0",
      "code_extract": "OpenSSL::Digest.digest(\\"sha1\\", user.password)"
    },
    {
      "cwe_ids": [
        "326"
      ],
      "id": "ruby_lang_weak_password_hash_sha",
      "title": "Usage of weak hashing library on a password (SHA)",
      "description": "## Description\\n\\nA weak hashing library can lead to data breaches and greater security risk.\\n\\n## Remediations\\nAccording to [OWASP](https://owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/09-Testing_for_Weak_Cryptography/04-Testing_for_Weak_Encryption), SHA and SHA1 are considered weak hash algorithms and therefore shouldn't be used.\\n\\n✅ Use a recommended hashing library such as bcrypt or Argon2id for passwords\\n\\n\`\`\`ruby\\nBCrypt::Password.create(\\"password_123\\")\\n\`\`\`\\n\\n\`\`\`ruby\\nhasher = Argon2::Password.new()\\nhasher.create(\\"password_123\\")\\n\`\`\`\\n\\n## Resources\\n- [OWASP Password Storage Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html)\\n- [BCrypt Explained](https://dev.to/sylviapap/bcrypt-explained-4k5c)\\n- [ruby-argon2](https://github.com/technion/ruby-argon2)\\n",
      "documentation_url": "https://docs.bearer.com/reference/rules/ruby_lang_weak_password_hash_sha",
      "line_number": 4,
      "full_filename": "/tmp/bearer-scan/openssl_sha1_data.rb",
      "filename": ".",
      "data_type": {
        "category_uuid": "dd88aee5-9d40-4ad2-8983-0c791ddec47c",
        "name": "Passwords"
      },
      "category_groups": [
        "PII",
        "Personal Data"
      ],
      "source": {
        "start": 4,
        "end": 4,
        "column": {
          "start": 10,
          "end": 23
        }
      },
      "sink": {
        "start": 4,
        "end": 4,
        "column": {
          "start": 1,
          "end": 24
        },
        "content": "y.digest(user.password)"
      },
      "parent_line_number": 4,
      "snippet": "y.digest(user.password)",
      "fingerprint": "a917eee0d88b45489c7189f37e092168_1",
      "old_fingerprint": "cd1923849b36c6435d5712d897d6da17_1",
      "code_extract": "y.digest(user.password)"
    }
  ]
}"
`;
